File: app/src/main/java/com/example/purepawapp/ui/product/ProductRowAdapter.java
package com.example.purepawapp.ui.product;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.purepawapp.data.model.Product;
import com.example.purepawapp.databinding.ItemProductRowBinding;
import com.example.purepawapp.util.CurrencyUtils;

import java.util.function.Consumer;

public class ProductRowAdapter extends ListAdapter<Product, ProductRowAdapter.ViewHolder> {

    private final Consumer<Product> onClick;

    private static final DiffUtil.ItemCallback<Product> DIFF_CALLBACK = new DiffUtil.ItemCallback<>() {
        @Override
        public boolean areItemsTheSame(@NonNull Product oldItem, @NonNull Product newItem) {
            return oldItem.getId().equals(newItem.getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull Product oldItem, @NonNull Product newItem) {
            return oldItem.equals(newItem);
        }
    };

    public ProductRowAdapter(Consumer<Product> onClick) {
        super(DIFF_CALLBACK);
        this.onClick = onClick;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemProductRowBinding binding = ItemProductRowBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private final ItemProductRowBinding binding;

        ViewHolder(ItemProductRowBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(Product product) {
            binding.tvIcon.setText(ProductUi.getEmoji(product));
            binding.tvName.setText(product.getName());
            binding.tvSubtitle.setText(product.getShortDescription());
            binding.tvRating.setText(String.format("⭐ %.1f", product.getAverageRating()));
            binding.tvPrice.setText(CurrencyUtils.toVndString(product.getDisplayPrice()));
            binding.getRoot().setOnClickListener(v -> onClick.accept(product));
        }
    }
}

File: app/src/main/java/com/example/purepawapp/ui/product/ProductRowAdapter.kt
package com.example.purepawapp.ui.product

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.purepawapp.data.model.Product
import com.example.purepawapp.databinding.ItemProductRowBinding
import com.example.purepawapp.util.toVndString

class ProductRowAdapter(
    private val onClick: (Product) -> Unit
) : ListAdapter<Product, ProductRowAdapter.ViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ItemProductRowBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class ViewHolder(private val binding: ItemProductRowBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(product: Product) {
            binding.tvIcon.text = product.emoji
            binding.tvName.text = product.name
            binding.tvSubtitle.text = product.shortDescription
            binding.tvRating.text = "⭐ ${"%.1f".format(product.averageRating)}"
            binding.tvPrice.text = product.displayPrice.toVndString()
            binding.root.setOnClickListener { onClick(product) }
        }
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<Product>() {
            override fun areItemsTheSame(oldItem: Product, newItem: Product): Boolean = oldItem.id == newItem.id
            override fun areContentsTheSame(oldItem: Product, newItem: Product): Boolean = oldItem == newItem
        }
    }
}
